File: packages/core/src/quality.ts
import type { Quality, StatsSample } from './types.js';

const HYSTERESIS = 2;

export function classify(sample: StatsSample): Quality {
  if (sample.loss > 0.1 || sample.rttMs > 400 || sample.jitterMs > 100) {
    return 'poor';
  }
  if (sample.loss > 0.03 || sample.rttMs > 200 || sample.jitterMs > 50) {
    return 'degraded';
  }
  return 'good';
}

export class QualityTracker {
  private label: Quality = 'good';
  private pending: Quality = 'good';
  private pendingCount = 0;
  private consecutivePoor = 0;

  observe(sample: StatsSample): { label: Quality; fallback: boolean } {
    const raw = classify(sample);

    if (raw === 'poor') this.consecutivePoor += 1;
    else this.consecutivePoor = 0;

    if (raw === this.label) {
      this.pending = raw;
      this.pendingCount = 0;
    } else if (raw === this.pending) {
      this.pendingCount += 1;
      if (this.pendingCount >= HYSTERESIS) {
        this.label = raw;
        this.pendingCount = 0;
      }
    } else {
      this.pending = raw;
      this.pendingCount = 1;
    }

    return { label: this.label, fallback: this.consecutivePoor >= 2 };
  }
}

File: android/engine/bin/main/io/sauti/engine/Quality.kt
package io.sauti.engine

private const val HYSTERESIS = 2

fun classifyQuality(sample: StatsSample): Quality {
    if (sample.loss > 0.1 || sample.rttMs > 400 || sample.jitterMs > 100) {
        return Quality.POOR
    }
    if (sample.loss > 0.03 || sample.rttMs > 200 || sample.jitterMs > 50) {
        return Quality.DEGRADED
    }
    return Quality.GOOD
}

data class QualityObservation(val label: Quality, val fallback: Boolean)

class QualityTracker {
    private var label: Quality = Quality.GOOD
    private var pending: Quality = Quality.GOOD
    private var pendingCount = 0
    private var consecutivePoor = 0

    fun observe(sample: StatsSample): QualityObservation {
        val raw = classifyQuality(sample)

        if (raw == Quality.POOR) consecutivePoor += 1 else consecutivePoor = 0

        when {
            raw == label -> {
                pending = raw
                pendingCount = 0
            }
            raw == pending -> {
                pendingCount += 1
                if (pendingCount >= HYSTERESIS) {
                    label = raw
                    pendingCount = 0
                }
            }
            else -> {
                pending = raw
                pendingCount = 1
            }
        }

        return QualityObservation(label, consecutivePoor >= 2)
    }
}

private val QUALITY_RANK = mapOf(
    Quality.GOOD to 0,
    Quality.DEGRADED to 1,
    Quality.POOR to 2
)

fun worstQuality(a: Quality, b: Quality): Quality =
    if (QUALITY_RANK.getValue(a) >= QUALITY_RANK.getValue(b)) a else b
